### PR TITLE
Enable static analysis

### PR DESCRIPTION
**Pull request by jibeee**
Add jobs for Clang Static Analyzer and CodeQL. Static analysis will be performed on each pull request and each commit to `master` and `develop`.

This PR also removes every compilation warnings, and the defects identified by Clang Static Analyzer and CodeQL.

### CODE REVIEW

File: src/monero_key.c [hunk 1]
```diff
@@ -215,8 +215,8 @@ int monero_apdu_display_address() {
     if (minor | major) {
         monero_get_subaddress(C, D, index);
     } else {
-        os_memmove(C, G_monero_vstate.A, 32);
-        os_memmove(D, G_monero_vstate.B, 32);
+        memcpy(C, G_monero_vstate.A, 32);
```
Comment (fbeutin-ledger) [inline src/monero_key.c:218]: Small remark, in cases like this how can we ensure that the value `C` computed does not overlap with `G_monero_vstate.A`

File: src/monero_key.c [hunk 2]
```diff
@@ -263,7 +263,7 @@ int monero_apdu_put_key() {
     monero_io_fetch(sec, 32);
     monero_io_fetch(pub, 32);
     monero_ecmul_G(raw, sec);
-    if (os_memcmp(pub, raw, 32)) {
+    if (memcmp(pub, raw, 32)) {
```
Comment (fbeutin-ledger) [inline src/monero_key.c:266]: ```if (memcmp(pub, raw, 32) != 0) {```

File: src/monero_key.c [hunk 3]
```diff
@@ -273,7 +273,7 @@ int monero_apdu_put_key() {
...
```
Comment (fbeutin-ledger) [inline src/monero_key.c:276]: ```if (memcmp(pub, raw, 32) != 0) {```

File: src/monero_prehash.c [hunk 4]
```diff
@@ -119,9 +119,9 @@ int monero_apdu_mlsag_prehash_update() {
             monero_ecmul_H(aH, v);
             monero_ecadd(aH, kG, aH);
         } else {
-            os_memmove(aH, kG, 32);
+            memcpy(aH, kG, 32);
         }
-        if (os_memcmp(C, aH, 32)) {
+        if (memcmp(C, aH, 32)) {
```
Comment (fbeutin-ledger) [inline src/monero_prehash.c:124]: ```if (memcmp(C, aH, 32) != 0) {```

File: src/monero_prehash.c [hunk 5]
```diff
@@ -131,7 +131,7 @@ int monero_apdu_mlsag_prehash_update() {
             if (G_monero_vstate.io_protocol_version >= 2) {
                 // finalize and check destination hash_control
                 monero_sha256_outkeys_final(k);
-                if (os_memcmp(k, G_monero_vstate.OUTK, 32)) {
+                if (memcmp(k, G_monero_vstate.OUTK, 32)) {
```
Comment (fbeutin-ledger) [inline src/monero_prehash.c:134]: `!= 0`

File: src/monero_prehash.c [hunk 6]
```diff
@@ -176,7 +176,7 @@ int monero_apdu_mlsag_prehash_finalize() {
         // Finalize and check commitment hash control
         if (G_monero_vstate.tx_sig_mode == TRANSACTION_CREATE_REAL) {
             monero_sha256_commitment_final(H);
-            if (os_memcmp(H, G_monero_vstate.C, 32)) {
+            if (memcmp(H, G_monero_vstate.C, 32)) {
```
Comment (fbeutin-ledger) [inline src/monero_prehash.c:179]: `!= 0`

File: src/monero_ux_nano.c [hunk 7]
```diff
@@ -350,11 +357,11 @@ void account_submenu_selector(unsigned int idx) {
     ui_menu_main_display(0);
 }
 
-void ui_menu_account_display(unsigned int value) {
+void ui_menu_account_display(void) {
     ux_menulist_init(G_ux.stack_count - 1, account_submenu_getter, account_submenu_selector);
 }
 
-void settings_change_account(void) { ui_menu_account_display(0); }
+void settings_change_account(void) { ui_menu_account_display(); }
```
Comment (fbeutin-ledger) [inline src/monero_ux_nano.c:364]: Maybe multiline

File: src/monero_ux_nano.c [hunk 8]
```diff
@@ -429,13 +436,13 @@ void network_submenu_selector(unsigned int idx) {
     ui_menu_main_display(0);
 }
 
-void ui_menu_network_display(unsigned int value) {
+void ui_menu_network_display(void) {
     ux_menulist_init(G_ux.stack_count - 1, network_submenu_getter, network_submenu_selector);
 }
 
-void settings_change_network(void) { ui_menu_network_display(0); }
+void settings_change_network(void) { ui_menu_network_display(); }
```
Comment (fbeutin-ledger) [inline src/monero_ux_nano.c:443]: Maybe multiline

File: src/monero_ux_nano.c [hunk 9]
```diff
@@ -458,9 +465,9 @@ UX_STEP_CB(ux_menu_reset_3_step, pb, ui_menu_reset_action(ACCEPT),
 
 UX_FLOW(ux_flow_reset, &ux_menu_reset_1_step, &ux_menu_reset_2_step, &ux_menu_reset_3_step);
 
-void ui_menu_reset_display(unsigned int value) { ux_flow_init(0, ux_flow_reset, 0); }
+void ui_menu_reset_display(void) { ux_flow_init(0, ux_flow_reset, 0); }
 
-void settings_reset(void) { ui_menu_reset_display(0); }
+void settings_reset(void) { ui_menu_reset_display(); }
```
Comment (fbeutin-ledger) [inline src/monero_ux_nano.c:470]: Same, can be made multi lines

File: src/monero_ux_nano.c [hunk 10]
```diff
@@ -549,7 +556,7 @@ UX_FLOW(ux_flow_about,
 #endif
         &ux_menu_about_2_step);
 
-void ui_menu_about_display(unsigned int value) { ux_flow_init(0, ux_flow_about, NULL); }
+void ui_menu_about_display(void) { ux_flow_init(0, ux_flow_about, NULL); }
```
Comment (fbeutin-ledger) [inline src/monero_ux_nano.c:559]: Same, can be made multi lines

File: src/monero_ux_nanos.c [hunk 11]
```diff
@@ -104,16 +104,16 @@ const bagl_element_t* ui_menu_words_preprocessor(const ux_menu_entry_t* entry,
...
 }
 
-void ui_menu_words_clear(unsigned int value) {
+void ui_menu_words_clear(unsigned int value __attribute__((unused))) {
     monero_clear_words();
     ui_menu_main_display(0);
 }
 
-void ui_menu_words_back(unsigned int value) { ui_menu_settings_display(1); }
+void ui_menu_words_back(unsigned int value __attribute__((unused))) { ui_menu_settings_display(1); }
```
Comment (fbeutin-ledger) [inline src/monero_ux_nanos.c:116]: Can be made multi lines

**Comment by fbeutin-ledger** [comment]
CodeQL warning in file tests/test_sig.py
```Import of 'ClientNotSupported' is not used.```
Could be fixed

File: src/monero_io.c [hunk 12]
```diff
@@ -76,15 +76,15 @@ void monero_io_hole(unsigned int sz) {
     if ((G_monero_vstate.io_length + sz) > MONERO_IO_BUFFER_LENGTH) {
         THROW(ERROR_IO_FULL);
     }
-    os_memmove(G_monero_vstate.io_buffer + G_monero_vstate.io_offset + sz,
-               G_monero_vstate.io_buffer + G_monero_vstate.io_offset,
-               G_monero_vstate.io_length - G_monero_vstate.io_offset);
+    memmove(G_monero_vstate.io_buffer + G_monero_vstate.io_offset + sz,
```
[the remark this answers is not in the document]
Comment (jibeee) [reply src/monero_io.c:79]: Yes, in this PR `memmove` has been used when source and destination buffers may overlap.

File: src/monero_io.c [hunk 13]
```diff
@@ -66,7 +66,7 @@ void monero_io_discard(int clear) {
     }
 }
 
-void monero_io_clear() { os_memset(G_monero_vstate.io_buffer, 0, MONERO_IO_BUFFER_LENGTH); }
+void monero_io_clear() { memset(G_monero_vstate.io_buffer, 0, MONERO_IO_BUFFER_LENGTH); }
```
[the remark this answers is not in the document]
Comment (jibeee) [reply src/monero_io.c:69]: Code has been formatted with clang-format, which uses the `.clang-format` file of the project.

If I format the code as you suggest, linter job will probably fail.

File: src/monero_key.c [hunk 14]
```diff
@@ -241,9 +241,9 @@ int monero_apdu_display_address() {
 /* ----------------------------------------------------------------------- */
 /* ---                                                                 --- */
 /* ----------------------------------------------------------------------- */
-int is_fake_view_key(unsigned char *s) { return os_memcmp(s, C_FAKE_SEC_VIEW_KEY, 32) == 0; }
+int is_fake_view_key(unsigned char *s) { return memcmp(s, C_FAKE_SEC_VIEW_KEY, 32) == 0; }
 
-int is_fake_spend_key(unsigned char *s) { return os_memcmp(s, C_FAKE_SEC_SPEND_KEY, 32) == 0; }
+int is_fake_spend_key(unsigned char *s) { return memcmp(s, C_FAKE_SEC_SPEND_KEY, 32) == 0; }
```
[the remark this answers is not in the document]
Comment (jibeee) [reply src/monero_key.c:246]: I agree, feel free to modify `.clang-format`.

File: src/monero_key.c (hunk 1 above)
Comment (jibeee) [reply src/monero_key.c:218]: `G_monero_vstate.A` is a 32-byte global array, while `C` and `D` are 32-byte local arrays. They cannot overlap. Or I miss something?

File: src/monero_io.c [hunk 15]
```diff
@@ -234,7 +236,7 @@ static void monero_io_verify_hmac_for(const unsigned char* buffer, int len,
         hmac[36] = 0;
     }
     cx_hmac_sha256(G_monero_vstate.hmac_key, 32, hmac, 37, hmac, 32);
-    if (os_memcmp(hmac, expected_hmac, 32)) {
+    if (memcmp(hmac, expected_hmac, 32)) {
```
[the remark this answers is not in the document]
Comment (jibeee) [reply src/monero_io.c:239]: Agree, will fix it.

**Comment by j-berman** [comment]
Chiming in with my 2 piconeros: the changes look fine to me generally (I didn't review it deeply). But, I think it would be more conservative to keep this PR on hold for a future release *after* the release for the coming hard fork ([August 13th, block 2,688,888](https://www.getmonero.org/2022/04/20/network-upgrade-july-2022.html)). The PR doesn't fix any current bugs AFAICT -- I'd say stick with the current code pattern for now, and ship this PR once ready after the fork.

**Comment by jibeee** [comment]
Agree, no need to rush for these changes.

**Comment by fbeutin-ledger** [comment]
lgtm
The branch will need to be rebased and merged before to resolve the conflict though

**Review by fbeutin-ledger — APPROVED** [review]

**Review by fbeutin-ledger — APPROVED** [review]

**Comment by tjulien-ledger** [comment]
Hello @j-berman , as the latest update has been released, do you agree we can merge this PR now?

**Pull request merged**